### PR TITLE
Add Stateful (Global) Metrics

### PR DESCRIPTION
**Pull request by brge17**
This is a follow up on issue: https://github.com/keras-team/keras/issues/8657.

In this PR we introduce a Stateful (Global) Metrics base class and a single example - True Positives. Instead of computing the metric in batch wise averages, we directly store the state inside the metric and use the value of the state.

Note: Tests to come, I just wanted feedback on the scaffolding - I'm expecting a lot of feedback/revisions. The default metrics test  (https://github.com/keras-team/keras/blob/master/tests/keras/metrics_test.py) does not make sense for this metric as the state is a single number, and it should not return an array tensor - even for a misshaped input.

### CODE REVIEW

**Comment by brge17** [comment]
@ozabluda @Kritikalcoder @ori226 @hadsed if any of you four would like to help get this through the final stretch... that would be awesome. At some point each of you four expressed interest in the functionality for Keras.

File: keras/metrics.py [hunk 1]
```diff
@@ -20,6 +21,134 @@
...
+
+    # Properties
+        threshold: the lower limit on y_pred that counts as a positive class prediction
+        state: the current state of the metric through the current batch
+    """
+
+    def __init__(self, threshold=None):
+        """Set the threshold and name the metric
+
+        # Keyword Arguments
```
Comment (Dref360) [inline keras/metrics.py]: Just `# Arguments` to be consistent

File: keras/metrics.py [hunk 2]
```diff
@@ -20,6 +21,134 @@
...
+    global_metrics = []
+    global_metric_names = []
+    for m in metrics:
+        if isinstance(m, GlobalMetric):
+            global_metrics.append(m)
+            global_metric_names.append(serialize(m))
+    return global_metrics, global_metric_names
+
+
+class TruePositives(GlobalMetric):
```
Comment (Dref360) [inline keras/metrics.py]: Should be renamed BinaryTruePositive otherwise, people will want to use it for multiples classes. And for images.
It should be clear that `y_true` and `y_pred` must be [?,2]

File: keras/engine/training.py [hunk 3]
```diff
@@ -1332,6 +1334,11 @@ def _test_loop(self, f, ins, batch_size=None, verbose=0, steps=None):
             and/or metrics). The attribute `model.metrics_names` will give you
             the display labels for the scalar outputs.
         """
+        metrics_module.reset_global_metrics(self.metrics)
+        _, global_metric_names = metrics_module.get_global_metrics(self.metrics)
```
Comment (Dref360) [inline keras/engine/training.py]: Wouldn't : 
```
global_metric_indices = [self.metrics_names.index(name) for name in global_metrics_names]
```
Be more readable? An isn't name already a string so `str(name)` is redundant?

File: keras/metrics.py [hunk 4]
```diff
@@ -20,6 +21,138 @@
...
+        threshold: the lower limit on y_pred that counts as a positive class prediction
+        state: the current state of the metric through the current batch
+    """
+
+    def __init__(self, threshold=None):
+        """Set the threshold and name the metric
+
+        # Keyword Arguments
+            threshold: the lower limit on y_pred that counts as a postive class prediction.
+                Defaults to 0.5
```
Comment (Dref360) [inline keras/metrics.py]: If the default is 0.5, then I would suggest to put `def __init__(self, threshold=0.5):`

File: keras/engine/training.py (hunk 3 above)
Comment (brge17) [reply keras/engine/training.py]: I wanted a reusable helper function. ```str(name)```should be redundant but I'm always weary of string comparison in Python.

File: keras/metrics.py [hunk 5]
```diff
@@ -20,6 +21,133 @@
 from .losses import poisson
 from .losses import cosine_proximity
 from .utils.generic_utils import deserialize_keras_object
+from .engine.topology import _to_snake_case
+
+
+class GlobalMetric(object):
```
Comment (fchollet) [inline keras/metrics.py]: Maybe `StatefulMetric`?

File: keras/metrics.py [hunk 6]
```diff
@@ -20,6 +21,133 @@
...
+    # Methods:
+        update_states(self. y_true, y_pred)
+        reset_states(self)
+    """
+    def __init__(self):
+        """Initialize the Global Metric and give it a name.
+        """
+        # Instance name is class name with underscores
+        cls_name = self.__class__.__name__
+        self.__name__ = _to_snake_case(cls_name)
```
[the remark this answers is not in the document]
Comment (brge17) [reply keras/metrics.py]:
```def serialize(metric):
    return metric.__name__```

```.__name__``` is required to work with the serializer

[the remark this answers is not in the document]
Comment (fchollet) [reply keras/metrics.py]: Then the serializer should be updated, and metrics classes should have a `get_config`/`from_config` method

File: keras/metrics.py (hunk 2 above)
Comment (brge17) [reply keras/metrics.py]: Good point.

File: keras/metrics.py [hunk 7]
```diff
@@ -20,6 +21,133 @@
...
+
+    def __call__(self, y_true, y_pred):
+        return self.update_states(y_true, y_pred)
```
[the remark this answers is not in the document]
Comment (brge17) [reply keras/metrics.py]: I incorporated the ``self.add_update`` change.

Related to the second point, I think metrics being derived from the ``layer`` class would be rather confusing/unintuitive to people implementing their own stateful metrics. Even if metrics was a subclass of ``layers`` they would still require custom handling inside the training loop because of the existence of both stateful and non-stateful metrics in ``model.metrics``.

File: keras/metrics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (brge17) [reply keras/metrics.py]: Ack. I'll add this to my todo list, along with writing lots of tests.

**Comment by brge17** [comment]
Slowly adding tests, still need to do the get_config, from_config

File: keras/metrics.py [hunk 8]
```diff
@@ -20,6 +21,140 @@
...
+        metrics: a list of metric instances.
+    """
+    if metrics is not None:
+        for metric in metrics:
+            if isinstance(metric, StatefulMetric):
+                metric.reset_states()
+
+
+def get_stateful_metrics(metrics):
+    """ Return a list of stateful metrics.
```
Comment (bfollinprm) [inline keras/metrics.py]: Update this docstring.

File: keras/engine/training.py (hunk 3 above)
Comment (brge17) [reply keras/engine/training.py]: Also worth noting, that it will not work with repeated metric names with that logic.

**Comment by brge17** [comment]
Yes, it can be reimplemented to inherit from layer and if it would be a deal breaker, I'm more than happy to do it that way.

Intuitively metrics and layers are different things. In my view, despite having similarities that are convenient for inheritance, they are conceptually different and thus should be independent of one another.

**Comment by fchollet** [comment]
The goal is to have consistent APIs and minimize code redundancy. It may be that the only code you need to write is the subclasses (such as `TruePositives`, etc).

**Comment by brge17** [comment]
The larger issue I'm trying to debug currently, is that it only works on the TensorFlow backend.

See below:

```
File "/Users/briandev/Desktop/keras-dev/keras/keras/models.py", line 827, in compile
    **kwargs)
  File "/Users/briandev/Desktop/keras-dev/keras/keras/engine/training.py", line 914, in compile
    handle_metrics(output_metrics)
  File "/Users/briandev/Desktop/keras-dev/keras/keras/engine/training.py", line 911, in handle_metrics
    mask=masks[i])
  File "/Users/briandev/Desktop/keras-dev/keras/keras/engine/training.py", line 444, in weighted
    return K.mean(score_array)
  File "/Users/briandev/Desktop/keras-dev/keras/keras/backend/theano_backend.py", line 549, in mean
    if 'int' in x.dtype or x.dtype == 'bool':
```

**Comment by brge17** [comment]
In Theano ```tensor = (/variable, Elemwise{add,no_inplace}.0)``` is returned by the metric, instead of the value in TensorFlow.

Full traceback:
```
tests/keras/metrics_test.py:121:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
keras/backend/theano_backend.py:294: in eval
    return to_dense(x).eval()
keras/backend/theano_backend.py:92: in to_dense
    if is_sparse(tensor):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

tensor = (/variable, Elemwise{add,no_inplace}.0)

    def is_sparse(tensor):
>       return th_sparse_module and isinstance(tensor.type, th_sparse_module.SparseType)
```
The call only uses K.backend functions...

**Comment by fchollet** [comment]
Does it work with CNTK?

**Comment by fchollet** [comment]
I'll have a direct look later today (i.e. I will try coding this up)

**Comment by brge17** [comment]
It only works with TensorFlow. Here was the driver code (even-odd mnist) I was playing with for fun:


```import theano
import sys
sys.path.append("/Users/briandev/Desktop/keras-dev/keras")

import numpy as np

import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras.metrics import TruePositives

batch_size = 1000
num_classes = 2
epochs = 2

# input image dimensions
img_rows, img_cols = 28, 28

# the data, shuffled and split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')[0:1000]/255.0
x_test = x_test.astype('float32')[0:1000]/255.0
y_train = keras.utils.to_categorical(y_train[0:1000] % 2, num_classes)
y_test = keras.utils.to_categorical(y_test[0:1000] % 2, num_classes)

print('x_train shape:', x_train.shape)
print('y_train shape:', y_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=[TruePositives(0.5), TruePositives(0.0)])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test),
          )

score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=0)

print score

y_hat_test = model.predict(x_test, verbose=0)
y_ht = y_hat_test[:, 1] > .5
y_t = y_test[:, 1] == 1
tp = np.sum(np.multiply(y_t[y_t==True], y_ht[y_t==True]))
print "True Positives: % f" % tp
print "Keras calculated True Positives: % f" % score[1]

```

**Comment by fchollet** [comment]
I wrote the following, which works with all backends and is a lot simpler:

```python
from .engine.topology import Layer

class StatefulMetric(Layer):  # Semantics
    pass


class TruePositives(StatefulMetric):
    """Stateful Metric to count the total true positives over all batches.

    # Properties
        threshold: the lower limit on y_pred that counts as a
            positive class prediction
        state: the current state of the metric through the current batch
    # Note
        y_true, y_pred must be [?, 2] tensors that are one-hot encoded.
    """

    def __init__(self, threshold=0.5):
        """Set the threshold and name the metric

        # Arguments
            threshold: the lower limit on y_pred that counts as a
                positive class prediction. Defaults to 0.5
        """
        super(TruePositives, self).__init__()

        self.threshold = K.variable(value=threshold)
        self.true_positives = K.variable(value=0, dtype='int32')

    def reset_states(self):
        """Reset the state at the beginning of training and evaluation for each
         epoch.
        """
        K.set_value(self.true_positives, 0)

    def __call__(self, y_true, y_pred):
        """Update the state at the completion of each batch.

        # Arguments
            y_true: the batch_wise labels
            y_pred: the batch_wise predictions

        # Returns
            The total number of true positives seen this epoch at the
                completion of the batch.
        """

        # Slice the positive score
        y_true = K.cast(y_true[:, 1], 'int32')
        y_pred = y_pred[:, 1]
        # Softmax -> probabilities
        # Hey, this is making a lot of assumptions...
        y_pred = K.cast(y_pred >= self.threshold, 'int32')
        # c = correct classifications
        c = K.cast(K.equal(y_pred, y_true), 'int32')
        # tp_batch = number of true positives in a batch
        tp_batch = K.sum(c * y_true)
        updated = self.true_positives + tp_batch
        self.add_update(K.update_add(self.true_positives, tp_batch),
                        inputs=[y_true, y_pred])
        return updated
```

Notes:
- Not happy with serialization and naming at all. This should be *much* simpler
- The way true positives are computed looks weird to me, it would only work in a special case.
- Counts should be integers.

**Comment by brge17** [comment]
I like it, cleans up a lot of the code.

The actual metric TruePositive is not great (I was going to clean it up), assumes a very specific structure of the input (2-class one hot encoding). The PR is more about supporting stateful metrics, than implementing specific stateful metrics.

How do you propose making the serialization and naming simpler?

**Comment by fchollet** [comment]
> How do you propose making the serialization and naming simpler?

Metrics should be fully serializable as a matter of principle. That means:

```python
def serialize(metric):
    return serialize_keras_object(metric)


def deserialize(name, custom_objects=None):
    return deserialize_keras_object(name,
                                    module_objects=globals(),
                                    custom_objects=custom_objects,
                                    printable_module_name='metric function')
```

Other than that I would recommend:

- have a `name` argument in `TruePositive` constructor with default `true_positives`
- refer to a stateful metric via `metric.name` (in `training.py`, etc)
- stateful metrics should implement `get_config`, just like other layers, to serialize custom arguments
- consider having a private method `_process_metric_name` to handle both functions and layers with a same method

e.g. 

```python

def _process_metric_name(metric):
    if hasattr(metric, 'name'):
        return metric.name
    else:
        return metric.__name__
```

(maybe not the best, but you get the idea)

**Comment by fchollet** [comment]
@brge17 would you like to keep working on it, or should I take over?

**Comment by brge17** [comment]
I'm pressed on cycles at my job, I would appreciate the help greatly :). Feel free to take it over.

**Comment by ozabluda** [comment]
Reposting my comments from the referenced Issue for the reference:

https://github.com/keras-team/keras/issues/8657#issuecomment-348659254
https://github.com/keras-team/keras/issues/8657#issuecomment-348705035

**Comment by fchollet** [comment]
See: https://github.com/keras-team/keras/pull/9253
